### PR TITLE
[5.4] Make with

### PR DESCRIPTION
**Pull request by taylorotwell**
This allows functionality similar to old `$container->make(class, params)` functionality.

### CODE REVIEW

**Comment by taylorotwell** [comment]
This still needs some work but is a proof of concept.

**Comment by taylorotwell** [comment]
I still find this whole feature quite odd... curious what the behavior should even be when resolving things that are singletons or instances that have already been resolved.

**Comment by taylorotwell** [comment]
All my tests pass now. If anyone sees any broken behavior it with it feel free to let me know so I can add a test.

File: src/Illuminate/Container/Container.php [hunk 1]
```diff
@@ -710,7 +710,7 @@ public function build($concrete)
         // hand back the results of the functions, which allows functions to be
         // used as resolvers for more fine-tuned resolution of these objects.
         if ($concrete instanceof Closure) {
-            return $concrete($this, last($this->with) ?: []);
+            return $concrete($this, last($this->with));
```
Comment (thecrypticace) [inline src/Illuminate/Container/Container.php]: Using `last` here means that the composer.json needs to be updated to include `illuminate/support`

Comment (taylorotwell) [inline src/Illuminate/Container/Container.php]: OK I will remove those calls.

**Comment by taylorotwell** [comment]
I should note this makes a change to the old behavior of the `make(abstract, params)` method. In this version, it will never use a previously resolved singleton instance when calling `makeWith`. In my opinion it should make a new instance each time this method is called because the given parameter array is dynamic.

**Comment by Daveawb** [comment]
Hey Taylor, thanks for this. I know you mentioned that back and forth on this is not helping and I do agree.

I have a small problem with this PR. The `make` method should have the same signature as `makeWith` (as it was), PSR-11 declares a get method, not a make method. Wouldn't it be more prudent to create a `get` method that takes a single parameter and just mark `make` as deprecated for removal in L6. At that point maybe alias `make` to `get` and finally implement a `makeWith` method at that point?

This would solve a lot of community issues up front without code changes (unless already running on 5.4), and implement the methods you need for PSR-11 compatibility in the future?

**Comment by taylorotwell** [comment]
Make can not have same signature as makeWith on 5.4 because we don't make breaking changes on point releases. Adding that parameter would break the interface.

I personally am totally apathetic towards PSR-11 as I think it's fairly useless overall but since it is so simple to add I don't have a huge problem with adding compatibility for it in Laravel 5.5.

**Comment by Daveawb** [comment]
Haha, and here I was, wrongly assuming that this was all gearing the container up for PSR-11. Thanks for clearing that up.

I'm sure it'd be BC if the signature became `make($abstract, $parameters = null)`?

Either way, nit picking and I'm grateful you've put this PR together.

**Comment by taylorotwell** [comment]
No, it wouldn't be backwards compatible. Look at the container interface.

**Comment by RemiCollin** [comment]
It wouldn't break the interface if `$parameters` is optionnal. @Daveawb  is right.

**Comment by devcircus** [comment]
Changing an existing public signature to add an optional param is a break. Might not affect many, but still a break.

**Comment by RemiCollin** [comment]
How would it affect them ? Even in the hypothesis some users were extending the base container without the optional parameter, it would still pass :

```php

interface ContainerInterface {

	public function make($abstract);
}

class Container implements ContainerInterface {

	public function make($abstract, $parameters = null) 
	{
		// works
	}
}


class ExtendedContainer extends Container {

	public function make($abstract)
	{
		// works as well
	}
}

```

**Comment by Daveawb** [comment]
@RemiCollin Actually looking at it, Taylor is absolutely correct when he says look at the interface for the container. Adding the extra parameter to the interface would not play well with any extended container instances that do not currently add an optional parameter. Alas, this is what it is. Looking forward to this PR being merged.

**Comment by RemiCollin** [comment]
Yes, if you change the `Container` interface. But that would not be mandatory in that case, if you look at my example. 

Mostly wanted to showcase how it would be possible to maintain pre-5.4 compatibility, while not introducing BC in current release. Not saying it would be a good pratice or not, it's totally debatable, but I think that any tool/language feature we have to provide users with smoother upgrades should be considered.

**Comment by taylorotwell** [comment]
People, I'm not going to argue if it is a breaking change or not. It obviously is if you understand how PHP works. Please no more discussion on that topic.

**Comment by taylorotwell** [comment]
Do people even want this feature or not. People were freaking out about wanting in the issue thread and now its just a couple of people arguing about BC. I'm not going to bother with it (because I think the feature is kinda silly) if this isn't actually a needed feature.

**Comment by RemiCollin** [comment]
Not arguing, just discussing implementation. 

But you're right, I won't bother if you feel it's an inappropriate discussion to have on a PR.

**Comment by Daveawb** [comment]
@taylorotwell Gosh the communitys wants and needs have really ground you down man. No arguments to be had here, just, I hope, constructive conversation. The community does absolutely want this, I'm not sure why no one else has chimed in and given a thumbs up but I speak for many people on my team who would very much like this functionality back.

**Comment by terion-name** [comment]
@taylorotwell With this it is impossible to pass a scalar in params :(

**Comment by saumya04** [comment]
@taylorotwell Thanks for the feature. I think you should mention this method `makeWith()` method on laravel's website under the changelog section.

**Comment by jonnywilliamson** [comment]
>Do people even want this feature or not?

I didn't know I wanted this feature until about 1 hour ago. Has make it so much easier to pass different configs into generating Guzzle clients now depending on if I'm testing/debugging/in production.

So thank you!

**Comment by benjoox** [comment]
> Do people even want this feature or not?

It saved my time too. Thanks!

**Comment by skovmand** [comment]
Thank you, thank you for this re-addition!

**Comment by antoniopaisfernandes** [comment]
Hi all,

I'm having an issue with testing/mocking/replacing the bind in the container.

Let me try to explain.

```$this->source->type``` is, for this test, LocalFolder
```$this->source``` is the source of data

This is the implementation:
```
$handler = app()->makeWith(
    $this->source->type,
    ['source' => $this->source]
);

$fileContents = $handler->get($item);
```

Everything working nicely in other tests but when I try to replace that class, it doesn't hit the replacement:

```
$this->app->bind(LocalFolder::class, new class($source) extends LocalFolder {
    public function get($item)
    {
        dd("IT HIT");
        return 'MOCKED_CONTENT';
    }
});
```

It never hits "IT HIT" :(

Do you know if there is any problem/issue with binding and resolving with ```makeWith```?

Thank you so much.
Best regards

**Comment by browner12** [comment]
this is not the appropriate place to ask for debugging help.  Please ask in the Slack or on a forum (stackoverflow or others).  Thanks

**Comment by giginos** [comment]
Got this behavior lost in Laravel 5.5 again?

**Pull request merged**